File: backend/app/schemas/mt4_diagnostics.py
from __future__ import annotations

from typing import Any

from pydantic import BaseModel, ConfigDict

from app.services.mt4_diagnostics import Mt4DiagnosticsResult


class Mt4DiagnosticsResponse(BaseModel):
    stage: str
    status_code: str
    data_quality_passed: bool
    can_proceed_to_read_only_analysis: bool
    is_tradable: bool
    note: str
    read_summary: dict[str, Any]
    metadata_status: dict[str, Any]
    freshness_status: dict[str, Any]
    gate_v0_result: dict[str, Any]
    required_fields_status: dict[str, Any]
    field_types_status: dict[str, Any]
    numeric_ranges_status: dict[str, Any]
    cross_field_status: dict[str, Any]
    gate_v1_result: dict[str, Any]

    model_config = ConfigDict(extra="forbid")


def mt4_diagnostics_response(
    result: Mt4DiagnosticsResult,
) -> Mt4DiagnosticsResponse:
    return Mt4DiagnosticsResponse(
        stage=result.stage,
        status_code=result.status_code,
        data_quality_passed=result.data_quality_passed,
        can_proceed_to_read_only_analysis=(
            result.can_proceed_to_read_only_analysis
        ),
        is_tradable=False,
        note=result.note,
        read_summary=_read_summary(result),
        metadata_status=_metadata_status(result),
        freshness_status=_freshness_status(result),
        gate_v0_result=_gate_v0_result(result),
        required_fields_status=_required_fields_status(result),
        field_types_status=_field_types_status(result),
        numeric_ranges_status=_numeric_ranges_status(result),
        cross_field_status=_cross_field_status(result),
        gate_v1_result=_gate_v1_result(result),
    )


def _read_summary(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    summary = result.read_summary
    return {
        "total_files": summary.total_files,
        "present_file_count": summary.present_file_count,
        "readable_object_count": summary.readable_object_count,
        "missing_files": summary.missing_files,
        "not_file_paths": summary.not_file_paths,
        "invalid_json_files": summary.invalid_json_files,
        "not_object_files": summary.not_object_files,
        "read_error_files": summary.read_error_files,
        "all_files_present": summary.all_files_present,
        "all_json_valid": summary.all_json_valid,
        "all_objects": summary.all_objects,
        "can_proceed_to_metadata_checks": summary.can_proceed_to_metadata_checks,
        "status_code": summary.status_code,
    }


def _metadata_status(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.metadata_status
    return {
        "all_metadata_present": status.all_metadata_present,
        "all_file_types_match": status.all_file_types_match,
        "all_sources_match": status.all_sources_match,
        "can_proceed_to_freshness_checks": (
            status.can_proceed_to_freshness_checks
        ),
        "status_code": status.status_code,
        "files": {
            "live_tick": _metadata_file(status.live_tick),
            "latest_bars": _metadata_file(status.latest_bars),
            "symbol_spec": _metadata_file(status.symbol_spec),
            "account_snapshot": _metadata_file(status.account_snapshot),
        },
    }


def _metadata_file(status: Any) -> dict[str, Any]:
    return {
        "file_name": status.file_name,
        "expected_file_type": status.expected_file_type,
        "actual_file_type": status.actual_file_type,
        "has_schema_version": status.has_schema_version,
        "has_file_type": status.has_file_type,
        "has_source": status.has_source,
        "has_generated_at": status.has_generated_at,
        "file_type_matches": status.file_type_matches,
        "source_matches": status.source_matches,
        "can_proceed_to_freshness_checks": (
            status.can_proceed_to_freshness_checks
        ),
        "status_code": status.status_code,
        "error_codes": status.error_codes,
    }


def _freshness_status(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.freshness_status
    return {
        "all_generated_at_parseable": status.all_generated_at_parseable,
        "all_files_fresh": status.all_files_fresh,
        "any_file_stale": status.any_file_stale,
        "any_file_from_future": status.any_file_from_future,
        "can_proceed_to_data_quality_gate": (
            status.can_proceed_to_data_quality_gate
        ),
        "status_code": status.status_code,
        "files": {
            "live_tick": _freshness_file(status.live_tick),
            "latest_bars": _freshness_file(status.latest_bars),
            "symbol_spec": _freshness_file(status.symbol_spec),
            "account_snapshot": _freshness_file(status.account_snapshot),
        },
    }


def _freshness_file(status: Any) -> dict[str, Any]:
    return {
        "file_name": status.file_name,
        "generated_at": status.generated_at,
        "age_seconds": status.age_seconds,
        "max_age_seconds": status.max_age_seconds,
        "max_future_skew_seconds": status.max_future_skew_seconds,
        "is_parseable": status.is_parseable,
        "is_fresh": status.is_fresh,
        "is_stale": status.is_stale,
        "is_from_future": status.is_from_future,
        "can_proceed_to_data_quality_gate": (
            status.can_proceed_to_data_quality_gate
        ),
        "status_code": status.status_code,
        "error_codes": status.error_codes,
    }


def _gate_v0_result(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.gate_v0_result
    return {
        "stage": status.stage,
        "status_code": status.status_code,
        "can_proceed_to_schema_checks": status.can_proceed_to_schema_checks,
        "blocked_by_read_status": status.blocked_by_read_status,
        "blocked_by_metadata": status.blocked_by_metadata,
        "blocked_by_freshness": status.blocked_by_freshness,
        "reasons": status.reasons,
        "source_read_status_code": status.source_read_status_code,
        "source_metadata_status_code": status.source_metadata_status_code,
        "source_freshness_status_code": status.source_freshness_status_code,
        "is_tradable": False,
        "note": status.note,
    }


def _required_fields_status(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.required_fields_status
    return {
        "all_required_fields_present": status.all_required_fields_present,
        "can_proceed_to_value_checks": status.can_proceed_to_value_checks,
        "status_code": status.status_code,
        "reasons": status.reasons,
        "is_tradable": False,
        "note": status.note,
        "files": {
            "live_tick": _required_fields_file(status.live_tick),
            "latest_bars": _required_fields_file(status.latest_bars),
            "symbol_spec": _required_fields_file(status.symbol_spec),
            "account_snapshot": _required_fields_file(status.account_snapshot),
        },
    }


def _required_fields_file(status: Any) -> dict[str, Any]:
    return {
        "file_name": status.file_name,
        "required_fields": status.required_fields,
        "present_fields": status.present_fields,
        "missing_fields": status.missing_fields,
        "null_fields": status.null_fields,
        "all_required_fields_present": status.all_required_fields_present,
        "can_proceed_to_value_checks": status.can_proceed_to_value_checks,
        "status_code": status.status_code,
        "error_codes": status.error_codes,
    }


def _field_types_status(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.field_types_status
    return {
        "all_field_types_valid": status.all_field_types_valid,
        "can_proceed_to_value_checks": status.can_proceed_to_value_checks,
        "status_code": status.status_code,
        "reasons": status.reasons,
        "is_tradable": False,
        "note": status.note,
        "files": {
            "live_tick": _field_types_file(status.live_tick),
            "latest_bars": _field_types_file(status.latest_bars),
            "symbol_spec": _field_types_file(status.symbol_spec),
            "account_snapshot": _field_types_file(status.account_snapshot),
        },
    }


def _field_types_file(status: Any) -> dict[str, Any]:
    return {
        "file_name": status.file_name,
        "expected_field_types": status.expected_field_types,
        "valid_type_fields": status.valid_type_fields,
        "wrong_type_fields": status.wrong_type_fields,
        "type_issues": [
            {
                "field_path": issue.field_path,
                "expected_type": issue.expected_type,
                "actual_type": issue.actual_type,
            }
            for issue in status.type_issues
        ],
        "all_field_types_valid": status.all_field_types_valid,
        "can_proceed_to_value_checks": status.can_proceed_to_value_checks,
        "status_code": status.status_code,
        "error_codes": status.error_codes,
    }


def _numeric_ranges_status(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.numeric_ranges_status
    return {
        "all_numeric_ranges_valid": status.all_numeric_ranges_valid,
        "can_proceed_to_cross_field_checks": (
            status.can_proceed_to_cross_field_checks
        ),
        "status_code": status.status_code,
        "reasons": status.reasons,
        "is_tradable": False,
        "note": status.note,
        "files": {
            "live_tick": _numeric_ranges_file(status.live_tick),
            "latest_bars": _numeric_ranges_file(status.latest_bars),
            "symbol_spec": _numeric_ranges_file(status.symbol_spec),
            "account_snapshot": _numeric_ranges_file(status.account_snapshot),
        },
    }


def _numeric_ranges_file(status: Any) -> dict[str, Any]:
    return {
        "file_name": status.file_name,
        "checked_fields": status.checked_fields,
        "invalid_fields": status.invalid_fields,
        "range_issues": [
            {
                "field_path": issue.field_path,
                "rule": issue.rule,
            }
            for issue in status.range_issues
        ],
        "all_numeric_ranges_valid": status.all_numeric_ranges_valid,
        "can_proceed_to_cross_field_checks": (
            status.can_proceed_to_cross_field_checks
        ),
        "status_code": status.status_code,
        "error_codes": status.error_codes,
    }


def _cross_field_status(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.cross_field_status
    return {
        "all_cross_field_checks_valid": status.all_cross_field_checks_valid,
        "can_proceed_to_data_quality_gate_finalization": (
            status.can_proceed_to_data_quality_gate_finalization
        ),
        "status_code": status.status_code,
        "reasons": status.reasons,
        "is_tradable": False,
        "note": status.note,
        "files": {
            "live_tick": _cross_field_file(status.live_tick),
            "latest_bars": _cross_field_file(status.latest_bars),
            "symbol_spec": _cross_field_file(status.symbol_spec),
            "account_snapshot": _cross_field_file(status.account_snapshot),
        },
    }


def _cross_field_file(status: Any) -> dict[str, Any]:
    return {
        "file_name": status.file_name,
        "checked_rules": status.checked_rules,
        "violated_rules": status.violated_rules,
        "cross_field_issues": [
            {
                "rule_name": issue.rule_name,
                "field_paths": issue.field_paths,
                "message": issue.message,
            }
            for issue in status.cross_field_issues
        ],
        "all_cross_field_checks_valid": status.all_cross_field_checks_valid,
        "can_proceed_to_data_quality_gate_finalization": (
            status.can_proceed_to_data_quality_gate_finalization
        ),
        "status_code": status.status_code,
        "error_codes": status.error_codes,
    }


def _gate_v1_result(result: Mt4DiagnosticsResult) -> dict[str, Any]:
    status = result.gate_v1_result
    return {
        "stage": status.stage,
        "status_code": status.status_code,
        "data_quality_passed": status.data_quality_passed,
        "can_proceed_to_read_only_analysis": (
            status.can_proceed_to_read_only_analysis
        ),
        "blocked_by_gate_v0": status.blocked_by_gate_v0,
        "blocked_by_required_fields": status.blocked_by_required_fields,
        "blocked_by_field_types": status.blocked_by_field_types,
        "blocked_by_numeric_ranges": status.blocked_by_numeric_ranges,
        "blocked_by_cross_field_checks": status.blocked_by_cross_field_checks,
        "reasons": status.reasons,
        "source_gate_v0_status_code": status.source_gate_v0_status_code,
        "source_required_fields_status_code": (
            status.source_required_fields_status_code
        ),
        "source_field_types_status_code": status.source_field_types_status_code,
        "source_numeric_ranges_status_code": (
            status.source_numeric_ranges_status_code
        ),
        "source_cross_field_status_code": status.source_cross_field_status_code,
        "is_tradable": False,
        "note": status.note,
    }

File: frontend/src/api/mt4Diagnostics.ts
import { apiGet } from "./client";

export type Mt4DiagnosticsFileKey =
  | "live_tick"
  | "latest_bars"
  | "symbol_spec"
  | "account_snapshot";

export type Mt4DiagnosticsFileMap<TFileSummary> = Record<
  Mt4DiagnosticsFileKey,
  TFileSummary
>;

export type Mt4ReadSummary = {
  total_files: number;
  present_file_count: number;
  readable_object_count: number;
  missing_files: string[];
  not_file_paths: string[];
  invalid_json_files: string[];
  not_object_files: string[];
  read_error_files: string[];
  all_files_present: boolean;
  all_json_valid: boolean;
  all_objects: boolean;
  can_proceed_to_metadata_checks: boolean;
  status_code: string;
};

export type Mt4MetadataFileSummary = {
  file_name: string;
  expected_file_type: string;
  actual_file_type: string | null;
  has_schema_version: boolean;
  has_file_type: boolean;
  has_source: boolean;
  has_generated_at: boolean;
  file_type_matches: boolean;
  source_matches: boolean;
  can_proceed_to_freshness_checks: boolean;
  status_code: string;
  error_codes: string[];
};

export type Mt4MetadataStatusSummary = {
  all_metadata_present: boolean;
  all_file_types_match: boolean;
  all_sources_match: boolean;
  can_proceed_to_freshness_checks: boolean;
  status_code: string;
  files: Mt4DiagnosticsFileMap<Mt4MetadataFileSummary>;
};

export type Mt4FreshnessFileSummary = {
  file_name: string;
  generated_at: string | null;
  age_seconds: number | null;
  max_age_seconds: number;
  max_future_skew_seconds: number;
  is_parseable: boolean;
  is_fresh: boolean;
  is_stale: boolean;
  is_from_future: boolean;
  can_proceed_to_data_quality_gate: boolean;
  status_code: string;
  error_codes: string[];
};

export type Mt4FreshnessStatusSummary = {
  all_generated_at_parseable: boolean;
  all_files_fresh: boolean;
  any_file_stale: boolean;
  any_file_from_future: boolean;
  can_proceed_to_data_quality_gate: boolean;
  status_code: string;
  files: Mt4DiagnosticsFileMap<Mt4FreshnessFileSummary>;
};

export type Mt4GateV0ResultSummary = {
  stage: string;
  status_code: string;
  can_proceed_to_schema_checks: boolean;
  blocked_by_read_status: boolean;
  blocked_by_metadata: boolean;
  blocked_by_freshness: boolean;
  reasons: string[];
  source_read_status_code: string | null;
  source_metadata_status_code: string | null;
  source_freshness_status_code: string | null;
  is_tradable: false;
  note: string;
};

export type Mt4RequiredFieldsFileSummary = {
  file_name: string;
  required_fields: string[];
  present_fields: string[];
  missing_fields: string[];
  null_fields: string[];
  all_required_fields_present: boolean;
  can_proceed_to_value_checks: boolean;
  status_code: string;
  error_codes: string[];
};

export type Mt4RequiredFieldsStatusSummary = {
  all_required_fields_present: boolean;
  can_proceed_to_value_checks: boolean;
  status_code: string;
  reasons: string[];
  is_tradable: false;
  note: string;
  files: Mt4DiagnosticsFileMap<Mt4RequiredFieldsFileSummary>;
};

export type Mt4FieldTypeIssueSummary = {
  field_path: string;
  expected_type: string;
  actual_type: string;
};

export type Mt4FieldTypesFileSummary = {
  file_name: string;
  expected_field_types: Record<string, string>;
  valid_type_fields: string[];
  wrong_type_fields: string[];
  type_issues: Mt4FieldTypeIssueSummary[];
  all_field_types_valid: boolean;
  can_proceed_to_value_checks: boolean;
  status_code: string;
  error_codes: string[];
};

export type Mt4FieldTypesStatusSummary = {
  all_field_types_valid: boolean;
  can_proceed_to_value_checks: boolean;
  status_code: string;
  reasons: string[];
  is_tradable: false;
  note: string;
  files: Mt4DiagnosticsFileMap<Mt4FieldTypesFileSummary>;
};

export type Mt4NumericRangeIssueSummary = {
  field_path: string;
  rule: string;
};

export type Mt4NumericRangesFileSummary = {
  file_name: string;
  checked_fields: string[];
  invalid_fields: string[];
  range_issues: Mt4NumericRangeIssueSummary[];
  all_numeric_ranges_valid: boolean;
  can_proceed_to_cross_field_checks: boolean;
  status_code: string;
  error_codes: string[];
};

export type Mt4NumericRangesStatusSummary = {
  all_numeric_ranges_valid: boolean;
  can_proceed_to_cross_field_checks: boolean;
  status_code: string;
  reasons: string[];
  is_tradable: false;
  note: string;
  files: Mt4DiagnosticsFileMap<Mt4NumericRangesFileSummary>;
};

export type Mt4CrossFieldIssueSummary = {
  rule_name: string;
  field_paths: string[];
  message: string;
};

export type Mt4CrossFieldFileSummary = {
  file_name: string;
  checked_rules: string[];
  violated_rules: string[];
  cross_field_issues: Mt4CrossFieldIssueSummary[];
  all_cross_field_checks_valid: boolean;
  can_proceed_to_data_quality_gate_finalization: boolean;
  status_code: string;
  error_codes: string[];
};

export type Mt4CrossFieldStatusSummary = {
  all_cross_field_checks_valid: boolean;
  can_proceed_to_data_quality_gate_finalization: boolean;
  status_code: string;
  reasons: string[];
  is_tradable: false;
  note: string;
  files: Mt4DiagnosticsFileMap<Mt4CrossFieldFileSummary>;
};

export type Mt4GateV1ResultSummary = {
  stage: string;
  status_code: string;
  data_quality_passed: boolean;
  can_proceed_to_read_only_analysis: boolean;
  blocked_by_gate_v0: boolean;
  blocked_by_required_fields: boolean;
  blocked_by_field_types: boolean;
  blocked_by_numeric_ranges: boolean;
  blocked_by_cross_field_checks: boolean;
  reasons: string[];
  source_gate_v0_status_code: string | null;
  source_required_fields_status_code: string | null;
  source_field_types_status_code: string | null;
  source_numeric_ranges_status_code: string | null;
  source_cross_field_status_code: string | null;
  is_tradable: false;
  note: string;
};

export type Mt4DiagnosticsResponse = {
  stage: string;
  status_code: string;
  data_quality_passed: boolean;
  can_proceed_to_read_only_analysis: boolean;
  is_tradable: false;
  note: string;
  read_summary: Mt4ReadSummary;
  metadata_status: Mt4MetadataStatusSummary;
  freshness_status: Mt4FreshnessStatusSummary;
  gate_v0_result: Mt4GateV0ResultSummary;
  required_fields_status: Mt4RequiredFieldsStatusSummary;
  field_types_status: Mt4FieldTypesStatusSummary;
  numeric_ranges_status: Mt4NumericRangesStatusSummary;
  cross_field_status: Mt4CrossFieldStatusSummary;
  gate_v1_result: Mt4GateV1ResultSummary;
};

export function getMt4Diagnostics(): Promise<Mt4DiagnosticsResponse> {
  return apiGet<Mt4DiagnosticsResponse>("/api/mt4/diagnostics");
}
